File: 188.BestTimetoBuyandSellStockIV/bottom_up_dp.js
/**
 * @param {number} k
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (k, prices) {
  const n = prices.length,
    dp = [...Array(n + 1)].map((e) =>
      [...Array(k + 1)].map((_) => Array(2).fill(0))
    );

  for (let i = n - 1; i >= 0; i--) {
    for (let j = 1; j <= k; j++) {
      for (let h = 0; h < 2; h++) {
        let doNothing = dp[i + 1][j][h],
          doSomething;
        if (h) {
          doSomething = prices[i] + dp[i + 1][j - 1][0];
        } else {
          doSomething = dp[i + 1][j][1] - prices[i];
        }
        dp[i][j][h] = Math.max(doNothing, doSomething);
      }
    }
  }
  return dp[0][k][0];
};

File: SegmentTree/309.BestTimetoBuyandSellStockwithCooldown/DESC.md/BottomUpDP.java
class BottomUpDP {
  public int maxProfit(int[] prices) {
    int[] MP = new int[prices.length + 2];

    for (int i = prices.length - 1; i >= 0; i--) {
      int C1 = 0;
      // Case 1). buy and sell the stock
      for (int sell = i + 1; sell < prices.length; sell++) {
        int profit = (prices[sell] - prices[i]) + MP[sell + 2];
        C1 = Math.max(profit, C1);
      }

      // Case 2). do no transaction with the stock p[i]
      int C2 = MP[i + 1];

      // wrap up the two cases
      MP[i] = Math.max(C1, C2);
    }
    return MP[0];
  }
}
